### PR TITLE
[MNG-6831] replace anonymous classes by lambdas

### PR DESCRIPTION
**Pull request by KroArtem**
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java [hunk 1]
```diff
@@ -105,18 +105,12 @@ public DefaultArtifactResolver()
...
-                {
-                    command.run();
-                }
-            };
+            executor = Runnable::run;
         }
         else
         {
             executor = new ThreadPoolExecutor( threads, threads, 3, TimeUnit.SECONDS,
-                                               new LinkedBlockingQueue<Runnable>(), new DaemonThreadCreator() );
```
Comment (michael-o) [inline maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: This should be separate PR.

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java [hunk 2]
```diff
@@ -259,7 +253,7 @@ public ArtifactResolutionResult resolveTransitively( Set<Artifact> artifacts, Ar
                                                              ArtifactNotFoundException
     {
         return resolveTransitively(
-                artifacts, originatingArtifact, Collections.<String, Artifact>emptyMap(), localRepository,
```
Comment (michael-o) [inline maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: This should be separate PR.

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java [hunk 3]
```diff
@@ -307,7 +301,7 @@ public ArtifactResolutionResult resolveTransitively( Set<Artifact> artifacts, Ar
...
```
Comment (michael-o) [inline maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: This should be separate PR.

**Comment by michael-o** [comment]
> 
> 
> I don't find this makes the code easier to read. I prefer the inner classes in most of these cases.

At some point, I do agree. But there are improvements which squash from 10 to one line. These are worth merging.

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java (hunk 1 above)
Comment (olamy) [reply maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: why? sounds a bit nitpicking

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java (hunk 2 above)
Comment (olamy) [reply maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: why? sounds a bit nitpicking

**Comment by rfscholte** [comment]
Regarding the lambda's in general: for those that have been programming Java for over 15 years (like me) having anonymous classes feels more comfortable. Reading lambda's takes time for us, because in our minds we keep trying to read them with inner classes.
For the generation that grew up with lambda's it is a nobrainer.
This is probably a good moment to let go the comfortzone. We know that while Maven was on Java 7 we missed new contributors providing patches because they couldn't use these Java 8 features.

Regarding refactoring in general: refactoring should be a constant factor. It is part of the process and IMO often doesn't deserve a separate issuenumber.

**Comment by michael-o** [comment]
> Regarding refactoring in general: refactoring should be a constant factor. It is part of the process and IMO often doesn't deserve a separate issuenumber.

I object that because every code change can introduce regressions.

**Comment by khmarbaise** [comment]
> > Regarding refactoring in general: refactoring should be a constant factor. It is part of the process and IMO often doesn't deserve a separate issuenumber.
> 
> I object that because every code change can introduce regressions.
I partially agree but that means in consequence our tests are not good enough...which means we have to enhance the tests we have or add new tests for those cases...

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java (hunk 1 above)
Comment (khmarbaise) [reply maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: No need for a separate PR...

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java (hunk 2 above)
Comment (khmarbaise) [reply maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: No need for a separate PR...

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java (hunk 3 above)
Comment (khmarbaise) [reply maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: No need for a separate PR...

**Comment by khmarbaise** [comment]
Let's see what our CI says....

**Comment by khmarbaise** [comment]
> Regarding the lambda's in general: for those that have been programming Java for over 15 years (like me) having anonymous classes feels more comfortable. Reading lambda's takes time for us, because in our minds we keep trying to read them with inner classes.
> For the generation that grew up with lambda's it is a nobrainer.

If you really have a problem to understand a Lambda you can simply use your IDE to convert the Lambda back to old style ...if you do this several times you will understand how they work and it will become more and more familiar to read Lambdas.

**Comment by michael-o** [comment]
> 
> 
> > Regarding the lambda's in general: for those that have been programming Java for over 15 years (like me) having anonymous classes feels more comfortable. Reading lambda's takes time for us, because in our minds we keep trying to read them with inner classes.
> > For the generation that grew up with lambda's it is a nobrainer.
> 
> If you really have a problem to understand a Lambda you can simply use your IDE to convert the Lambda back to old style ...if you do this several times you will understand how they work and it will become more and more familiar to read Lambdas.

I join Robert here. Funny thing is that I have used lambdas way before I have used Java, namely with Haskell. It was a breeze to use.

**Comment by khmarbaise** [comment]
@KroArtem Hi, can you please rebase your PR's against current master and please squash your commits...

File: maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java (hunk 1 above)
Comment (KroArtem) [reply maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: Please decide whether to make a separate PR or not 😃

Comment (michael-o) [reply maven-compat/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactResolver.java]: Let's leave as-is. In the future, try to avoid unrelated changes.

**Comment by KroArtem** [comment]
@khmarbaise , hello, I've rebased the PR and squashed it. @michael-o , it seems like you're still requesting a review.

File: maven-settings-builder/src/main/java/org/apache/maven/settings/building/DefaultSettingsBuilder.java [hunk 4]
```diff
@@ -241,20 +240,17 @@ private Settings interpolate( Settings settings, SettingsBuildingRequest request
...
             {
-                if ( value != null )
-                {
-                    // we're going to parse this back in as XML so we need to escape XML markup
-                    value = value.toString().replace( "&", "&amp;" ).replace( "<", "&lt;" ).replace( ">", "&gt;" );
-                    return value;
-                }
-                return null;
+                // we're going to parse this back in as XML so we need to escape XML markup
+                value = value.toString().replace( "&", "&amp;" ).replace( "<", "&lt;" ).replace( ">", "&gt;" );
```
Comment (elharo) [inline maven-settings-builder/src/main/java/org/apache/maven/settings/building/DefaultSettingsBuilder.java]: not changed in this PR, but dollars to donuts this conversion is buggy

Comment (KroArtem) [reply maven-settings-builder/src/main/java/org/apache/maven/settings/building/DefaultSettingsBuilder.java]: I had to google `dollars to donuts` to understand your comment :)

Do you mean not only these conversions should be done? https://stackoverflow.com/questions/1091945/what-characters-do-i-need-to-escape-in-xml-documents

**Comment by KroArtem** [comment]
Are there any more things to fix here? If not, is it possible for someone to merge this?

File: maven-settings-builder/src/main/java/org/apache/maven/settings/building/DefaultSettingsBuilder.java (hunk 4 above)
Comment (elharo) [reply maven-settings-builder/src/main/java/org/apache/maven/settings/building/DefaultSettingsBuilder.java]: Nothing specific and not this PR's problem to fix. It just happened to call my attention to it. I've just learned from experience that these sorts of replacements always go wrong. Just a note to myself to dig into this, really.
